build.sbt

```
@@ -255,6 +295,8 @@ lazy val V = new {
   def nonDeprecatedScala3Versions = Seq(nextScala3RC, scala3, "3.0.2")
   def deprecatedScala3Versions = Seq("3.0.1", "3.0.0")
   def scala3Versions = nonDeprecatedScala3Versions ++ deprecatedScala3Versions
+  lazy val nightlyScala3Versions =
+    Scala3NightlyVersions.nightlyReleasesAfter(nextScala3RC)
```

```suggestion
  lazy val nightlyScala3Versions = { if (isCi)
    Scala3NightlyVersions.nightlyReleasesAfter(nextScala3RC)
    else Nil
  }
```

To avoid fetching the versions when starting sbt locally, which would add to load time. I would even add a separate variable so that this is only added when needed in the nightlies cross tests. Otherwise, each job will try to download the versions.

So maybe:
```suggestion
  lazy val nightlyScala3Versions = { if (System.getenv("NIGHTLIES") != null)
    Scala3NightlyVersions.nightlyReleasesAfter(nextScala3RC)
    else Nil
  }
```
and set `NIGHTLIES` for the single cross test and for the releases.

---

At the beginning I was thinking about additional costs from versions fetching but I was surprised by how fast it works an decided that probably it doesn't worth to prevent this call.

Just checked it more time: it takes less than 500ms.  Tested it with and without internet connection.
Maybe let's leave it as it is?

---

> it takes less than 500ms.

The problem is that at some point it all accumulates and sbt just loads super slow.
I would still change it to at least only run on CI, it's simple enough if anyone needs to make it work locally.

---

:smiling_face_with_tear: 
I understand your concerns but I think it will be confusing that you need to pass a special env to work with nightlies locally (test/debug/publish local). It also should be documented for contributors. 

Wdyt if I cover this thing by sbt cache? Probably it will be more faster.
The worst case will be that once a day coursier downloads this file(https://repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/maven-metadata.xml)   that I think should be about 500ms.

---

> I understand your concerns but I think it will be confusing that you need to pass a special env to work with nightlies locally (test/debug/publish local). It also should be documented for contributors.

To publish a nightly version all you need to do is `++3.0.1-RC1-bin-20210413-f3c1468-NIGHTLY` and `mtags/publishLocal`, so this wouldn't be a problem for releasing locally really. And with these changes when someone does `+publishLocal` they will have 5 more versions to publish each time even if everything was compiled last time.

I would really opt that it's all done in CI only.

---

Changed it - now it's needed  to set either `CI` or `NIGHTLIES` env var to enable this logic.